### PR TITLE
[MM-55009] Remove pre-fetch preference and set new pre fetch limits

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
The goal of this is to find a balance for pre-fetching and making the webapp more usable for people with a high number of unread channels. After experimenting on Community we came up with these limits.

1. <10 unread channels. Prefetch everything.

2. 10-20 unread. Prefetch only mentions, capped to 10.

3. ">20 unread. Don't prefetch anything.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55009

#### Release Note

```release-note
Remove pre-fetch preference and set new pre fetch limits for the webapp
```

### CODE REVIEW

File: webapp/channels/src/components/data_prefetch/index.ts [hunk 1]
```diff
@@ -39,39 +37,56 @@ enum Priority {
 // function to return a queue obj with priotiy as key and array of channelIds as values.
 // high priority has channels with mentions
 // medium priority has channels with unreads
+// <10 unread channels. Prefetch everything.
+// 10-20 unread. Prefetch only mentions, capped to 10.
+// >20 unread. Don't prefetch anything.
 const prefetchQueue = memoizeResult((
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/data_prefetch/index.ts:43]: Would you be open to adding a few unit tests for the conditions mentioned here

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/data_prefetch/index.ts [hunk 2]
```diff
@@ -39,39 +37,56 @@ enum Priority {
...
+        [Priority.low]: [], //TODO: add chanenls such as fav.
+    };
+    if (!unreadChannelsCount || unreadChannelsCount > 20) {
+        return defaultResult;
+    }
+    for (const channel of unreadChannels) {
         const channelId = channel.id;
         const membership = memberships[channelId];
 
+        if (unreadChannelsCount >= 10 && defaultResult[Priority.high].length >= 10) {
```
Comment (hmhealey) [inline webapp/channels/src/components/data_prefetch/index.ts:65]: Would you mind adding constants for the 10/20? It'll make it easier to update those if we want to tweak them a bit

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/data_prefetch/index.ts [hunk 3]
```diff
@@ -39,39 +37,56 @@ enum Priority {
...
+    let defaultResult: {
+        1: string[];
+        2: string[];
+        3: string[];
+    } = {
+        [Priority.high]: [], // 1 being high priority requests
+        [Priority.medium]: [],
+        [Priority.low]: [], //TODO: add chanenls such as fav.
+    };
+    if (!unreadChannelsCount || unreadChannelsCount > 20) {
```
[the remark this answers is not in the document]
Comment (BenCookie95) [reply webapp/channels/src/components/data_prefetch/index.ts:58]: Yeah there is a low chance of pre-loading the correct channels but you're right that it's odd. We could do the same behaviour as 10-20 where we pre-load your first 10 mentions?

[the remark this answers is not in the document]
Comment (agnivade) [reply webapp/channels/src/components/data_prefetch/index.ts:58]: The 10-20 range is a rough number that I came up with after asking several folks on community on their avg unread count. The reasoning is that when there are >20 unreads, the user behavior is that they don't actually read all the unreads, but rather go through them in batches. Therefore, any amount of preloading is actually wasted. Therefore, we decided not to preload anything.

**Review by M-ZubairAhmed — APPROVED** [review]

File: webapp/channels/src/components/data_prefetch/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmhealey) [reply webapp/channels/src/components/data_prefetch/index.ts]: Makes sense. This looks good then

**Review by hmhealey — APPROVED** [review]

**Pull request merged**